### PR TITLE
[core] limit parallelly read file memory usage, extract some methods

### PR DESCRIPTION
**Pull request by leaves12138**
[core] limit parallelly read file memory usage, extract some methods(#1061)

### Tests

*(List UT and IT cases to verify this change)*

### API and Format 

*(Does this change affect API or storage format)*

### Documentation

*(Does this change introduce a new feature)*

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 1]
```diff
@@ -195,6 +196,70 @@ public FileStoreScan withManifestCacheFilter(ManifestCacheFilter manifestFilter)
...
+                                                partInfo, numOfBuckets, file.totalBuckets()));
+                            }
+
+                            // bucket filter should not be applied along with partition filter
+                            // because the specifiedBucket is computed against the current
+                            // numOfBuckets
+                            // however entry.bucket() was computed against the old numOfBuckets
+                            // and thus the filtered manifest entries might be empty
+                            // which renders the bucket check invalid
+                            if (filterByBucket(file)
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:236]: 'if' statement can be simplified

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 2]
```diff
@@ -195,6 +196,70 @@ public FileStoreScan withManifestCacheFilter(ManifestCacheFilter manifestFilter)
...
+
+            @Override
+            public List<ManifestEntry> files() {
+                return files;
+            }
+        };
+    }
+
+    private <T extends AbstractManifestEntry> Pair<Long, List<T>> doPlan(
+            Function<List<ManifestFileMeta>, List<T>> processor,
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:261]: `readManifestFile`?

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 3]
```diff
@@ -195,6 +196,70 @@ public FileStoreScan withManifestCacheFilter(ManifestCacheFilter manifestFilter)
 
     @Override
     public Plan plan() {
+        Pair<Long, List<ManifestEntry>> planResult =
+                doPlan(
+                        // how to process entry files
+                        entries ->
+                                entries.parallelStream()
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:203]: Can we also put these logical into `doPlan`?
Only `filterByStats` is special, we can just `instanceof ManifestEntry` in `filterByStats`?

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 4]
```diff
@@ -195,6 +196,70 @@ public FileStoreScan withManifestCacheFilter(ManifestCacheFilter manifestFilter)
...
+            @Override
+            public List<ManifestEntry> files() {
+                return files;
+            }
+        };
+    }
+
+    private <T extends AbstractManifestEntry> Pair<Long, List<T>> doPlan(
+            Function<List<ManifestFileMeta>, List<T>> processor,
+            Function<T, Boolean> postFilterProcessor) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:262]: Maybe we don't need this one, postFilter can be applied to `SimpleManifestEntry` too.
If there is level filter, throw exception.

**Comment by leaves12138** [comment]
fixed comment

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * This class is a parallel execution util class, which mainly aim to process tasks parallelly with
+ * memory control.
+ */
+public class ParallellyExecuteUtils {
+
+    // reduce memory usage by batch iterable process, the cached result in memory will be 2 *
+    // queueSize
+    public static <T, U> Iterable<T> parallelismBatchIterable(
+            Function<List<U>, List<T>> processor, List<U> input, int queueSize) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:38]: default value of queueSize can be COMMON_IO_FORK_JOIN_POOL thread number * 2.

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                                                    () -> processor.apply(stack.poll()),
+                                                    FileUtils.COMMON_IO_FORK_JOIN_POOL);
+                                } else {
+                                    batch = null;
+                                }
+                            }
+                            if (index < activeList.size()) {
+                                next = activeList.get(index++);
+                                if (index == activeList.size()) {
+                                    activeList = null;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:94]: You implement a wrong iterator.
Test should cover multiple invoking for `hasNext `.

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                            // get active list for polling, and submit tasks meanwhile
+                            if (activeList == null) {
+                                if (batch == null) {
+                                    return false;
+                                }
+
+                                activeList = batch.get();
+                                if (stack.size() > 0) {
+                                    batch =
+                                            CompletableFuture.supplyAsync(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:84]: We don't need to produce more elements when the consumer not finish this batch.

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java (hunk 6 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java]: done this. added tests for multiple invoking hasNext method

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java (hunk 5 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java]: done this! set queueSize to FileUtils.COMMON_IO_FORK_JOIN_POOL.getParallelism() * 2

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 8]
```diff
@@ -316,19 +321,28 @@ private boolean filterManifestFileMeta(ManifestFileMeta manifest) {
...
+    private boolean filterByLevel(AbstractManifestEntry entry) {
+        return (levelFilter == null || levelFilter.test(entry.level()));
+    }
+
+    /** Note: Keep this thread-safe. */
+    private boolean filterByStats(AbstractManifestEntry entry) {
+        if (entry instanceof ManifestEntry) {
+            return filterByStats((ManifestEntry) entry);
+        } else {
+            throw new RuntimeException("only complete manifest entry could be filter by stats");
```
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java]: done this

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        if (queueSize <= 0) {
+            throw new NegativeArraySizeException("queue size should not be negetive");
+        }
+
+        // apart queue
+        final Queue<List<U>> stack = new ArrayDeque<>();
+        int size = input.size();
+        int num = size / queueSize;
+
+        for (int i = 0; i < num; i++) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:56]: `Lists.partition`?

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 10]
```diff
@@ -316,19 +320,27 @@ private boolean filterManifestFileMeta(ManifestFileMeta manifest) {
...
+    private boolean filterByLevel(AbstractManifestEntry entry) {
+        return (levelFilter == null || levelFilter.test(entry.level()));
+    }
+
+    /** Note: Keep this thread-safe. */
+    private boolean filterByStats(AbstractManifestEntry entry) {
+        if (entry instanceof ManifestEntry) {
+            return filterByStats((ManifestEntry) entry);
+        }
+        return true;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:343]: Add comments here: `filterByStats is an action that is completed as much as possible and does not have an impact if it is not done.`

Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:345]: done

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java (hunk 9 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java]: done

**Comment by leaves12138** [comment]
fixed comment

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * This class is a parallel execution util class, which mainly aim to process tasks parallelly with
+ * memory control.
+ */
+public class ParallellyExecuteUtils {
+
+    public static <T, U> Iterable<T> parallelismBatchIterable(
+            Function<List<U>, List<T>> processor, List<U> input) {
+        // default queueSize
+        int queueSize = FileUtils.COMMON_IO_FORK_JOIN_POOL.getParallelism() * 2;
```
Comment (wxplovecc) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:41]: Maybe default queueSize should be smaller , After testing it in the outdoor environment 
企业微信截图_16832893498454

Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:41]: Introduce an option: `SCAN_MANIFEST_PARALLELISM`, default is none (will be `COMMON_IO_FORK_JOIN_POOL.getParallelism()`).

Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java]: done this

**Comment by leaves12138** [comment]
fixed comment

**Comment by leaves12138** [comment]
test passed in my local environment?

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                                && stack.size() > 0) {
+                            // reset index
+                            index = 0;
+                            try {
+                                activeList =
+                                        CompletableFuture.supplyAsync(
+                                                        () -> processor.apply(stack.poll()),
+                                                        FileUtils.COMMON_IO_FORK_JOIN_POOL)
+                                                .get();
+                            } catch (InterruptedException e) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:80]: Just
```
} catch (Exception e) {
    throw new RuntimeException(e);
}
```
?

- Swallow `InterruptedException` is not good.
- Exception message is confused, `"should never get here"`, IOException will go here.

File: paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java [hunk 13]
```diff
@@ -66,7 +67,7 @@ public T next() {
                     }
 
                     private void advanceIfNeeded() {
-                        if ((activeList == null || index >= activeList.size())
+                        while ((activeList == null || index >= activeList.size())
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/ParallellyExecuteUtils.java:69]: Add a test for this?

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**